### PR TITLE
(SERVER-3167) Take up JRuby 9.4

### PR DESCRIPTION
**Pull request by justinstoller**

### CODE REVIEW

**Comment by justinstoller** [comment]
Need to validate a hack around JRuby vs Timeout thread termination.

**Comment by justinstoller** [comment]
fwiw, the tests are failing because we need a version of Puppet that pins the concurrent gem.

The change to Pupppet::Server::Master#terminate feels like a cludge that will introduce bugs later in the real world. I'm currently investigating other approaches while I wait for a response from the JRuby team.

**Comment by donoghuc** [comment]
I would imagine the puppet/concurrent issue will go away now that https://github.com/puppetlabs/puppetserver/commit/297a550f625e82f0fe4e8abac369fb0c725a29d6 is in

**Comment by justinstoller** [comment]
Some performance metrics of 9.4 vs 9.3 ran over the last couple of days using `bundle exec rake benchmark:full_catalog` in the Puppet repo. For our uses JRuby is slower but probably not so slow that it requires us to rethink how we're doing things:

9.3 Runs
========
```
➜  cat 93-run 
JRuby 9.3.0.0 ITERATIONS=10
  6.920000   0.060000   6.980000 (  2.574555)
  6.070000   0.040000   6.110000 (  2.473492)
  4.960000   0.020000   4.980000 (  2.334820)
  3.720000   0.040000   3.760000 (  2.135919)
  2.640000   0.050000   2.690000 (  1.930228)
> total:    73.490000   0.520000  74.010000 ( 28.410209)
> avg:       7.349000   0.052000   7.401000 (  2.841021)

JRuby 9.3.0.0 ITERATIONS=10
  6.580000   0.070000   6.650000 (  2.569545)
  4.830000   0.060000   4.890000 (  2.319469)
  4.830000   0.040000   4.870000 (  2.214397)
  4.200000   0.070000   4.270000 (  2.078084)
  4.420000   0.050000   4.470000 (  2.176139)
  2.760000   0.020000   2.780000 (  1.864392)
> total:    63.510000   0.640000  64.150000 ( 25.311698)
> avg:       6.351000   0.064000   6.415000 (  2.531170)

JRuby 9.3.0.0 ITERATIONS=100
  2.310000   0.020000   2.330000 (  1.941666)
  2.350000   0.030000   2.380000 (  1.863831)
  2.300000   0.030000   2.330000 (  1.909738)
  1.850000   0.020000   1.870000 (  1.719980)
  1.990000   0.010000   2.000000 (  1.466067)
  1.400000   0.010000   1.410000 (  1.280951)
> total:   259.870000   2.700000 262.570000 (171.522146)
> avg:       2.598700   0.027000   2.625700 (  1.715221)

JRuby 9.3.0.0 ITERATIONS=100
  1.340000   0.020000   1.360000 (  1.281822)
  1.810000   0.010000   1.820000 (  1.501614)
  1.660000   0.030000   1.690000 (  1.305674)
  1.630000   0.030000   1.660000 (  1.361664)
  1.390000   0.010000   1.400000 (  1.264001)
  1.700000   0.000000   1.700000 (  1.307373)
  1.380000   0.030000   1.410000 (  1.262353)
> total:   235.600000   2.630000 238.230000 (154.023512)
> avg:       2.356000   0.026300   2.382300 (  1.540235)

JRuby 9.3.0.0 ITERATIONS=200
  1.640000   0.020000   1.660000 (  1.272720)
  1.270000   0.010000   1.280000 (  1.190541)
  1.280000   0.010000   1.290000 (  1.185085)
  1.280000   0.010000   1.290000 (  1.185857)
  1.290000   0.010000   1.300000 (  1.204939)
  1.390000   0.010000   1.400000 (  1.207486)
  1.280000   0.010000   1.290000 (  1.187869)
> total:   366.030000   3.620000 369.650000 (271.330916)
> avg:       1.830150   0.018100   1.848250 (  1.356655)

JRuby 9.3.0.0 ITERATIONS=200
  1.310000   0.010000   1.320000 (  1.222551)
  1.310000   0.010000   1.320000 (  1.215762)
  1.420000   0.010000   1.430000 (  1.220744)
  1.310000   0.010000   1.320000 (  1.235382)
  1.400000   0.000000   1.400000 (  1.235143)
  1.320000   0.010000   1.330000 (  1.219120)
  1.320000   0.010000   1.330000 (  1.222566)
> total:   389.750000   4.030000 393.780000 (278.851202)
> avg:       1.948750   0.020150   1.968900 (  1.394256)
```

9.4 Runs
=========
```
Using JRuby 9.4.0.0 with ITERATIONS=10
  6.490000   0.060000   6.550000 (  2.753544)
  7.050000   0.030000   7.080000 (  2.709455)
  5.520000   0.050000   5.570000 (  2.463327)
  5.540000   0.030000   5.570000 (  2.325315)
  4.180000   0.050000   4.230000 (  2.211034)
  3.080000   0.030000   3.110000 (  2.010518)
> total:    65.280000   0.510000  65.790000 ( 26.772906)
> avg:       6.528000   0.051000   6.579000 (  2.677291)

Using JRuby 9.4.0.0 with ITERATIONS=10
  7.190000   0.070000   7.260000 (  2.633891)
  5.020000   0.020000   5.040000 (  2.453391)
  5.780000   0.040000   5.820000 (  2.443005)
  5.140000   0.060000   5.200000 (  2.313908)
  4.990000   0.040000   5.030000 (  2.273856)
  3.910000   0.030000   3.940000 (  2.119499)
  2.660000   0.040000   2.700000 (  1.921460)
> total:    61.030000   0.570000  61.600000 ( 25.369025)
> avg:       6.103000   0.057000   6.160000 (  2.536902)

Using JRuby 9.4.0.0 with ITERATIONS=100
  1.780000   0.020000   1.800000 (  1.445617)
  1.720000   0.020000   1.740000 (  1.445325)
  1.710000   0.020000   1.730000 (  1.437807)
  1.810000   0.010000   1.820000 (  1.471072)
  2.200000   0.010000   2.210000 (  1.493955)
  1.650000   0.010000   1.660000 (  1.420614)
> total:   281.400000   2.590000 283.990000 (172.411136)
> avg:       2.814000   0.025900   2.839900 (  1.724111)

Using JRuby 9.4.0.0 with ITERATIONS=100
  1.670000   0.010000   1.680000 (  1.350717)
  1.600000   0.020000   1.620000 (  1.346538)
  1.540000   0.010000   1.550000 (  1.335194)
  1.560000   0.010000   1.570000 (  1.340409)
  1.750000   0.040000   1.790000 (  1.391608)
  1.960000   0.010000   1.970000 (  1.459116)
  1.470000   0.020000   1.490000 (  1.352234)
> total:   265.840000   2.480000 268.320000 (164.066685)
> avg:       2.658400   0.024800   2.683200 (  1.640667)

Using JRuby 9.4.0.0 with ITERATIONS=200
  1.500000   0.030000   1.530000 (  1.333559)
  1.400000   0.010000   1.410000 (  1.286063)
  1.530000   0.010000   1.540000 (  1.312355)
  1.410000   0.010000   1.420000 (  1.303853)
  1.440000   0.000000   1.440000 (  1.300142)
  1.430000   0.000000   1.430000 (  1.291432)
  1.300000   0.020000   1.320000 (  1.287605)
> total:   411.250000   3.900000 415.150000 (294.821859)
> avg:       2.056250   0.019500   2.075750 (  1.474109)

Using JRuby 9.4.0.0 with ITERATIONS=200
  1.670000   0.000000   1.670000 (  1.335416)
  1.480000   0.010000   1.490000 (  1.304327)
  1.370000   0.020000   1.390000 (  1.284459)
  1.400000   0.010000   1.410000 (  1.288679)
  1.410000   0.010000   1.420000 (  1.282646)
  1.300000   0.010000   1.310000 (  1.270194)
  1.370000   0.020000   1.390000 (  1.267890)
> total:   391.950000   4.120000 396.070000 (290.260752)
> avg:       1.959750   0.020600   1.980350 (  1.451304)
```

**Comment by justinstoller** [comment]
Closed/Reopened to try and get Travis to test with the concurrent ruby bump

**Comment by donoghuc** [comment]
woah did we hit this bug in jruby https://github.com/jruby/jruby/issues/7572 ? https://github.com/puppetlabs/puppet/blob/f25f20da81a471a2744b843590e91e002e3ed164/lib/puppet/pal/pal_impl.rb#L279 https://github.com/puppetlabs/puppetserver/blob/297a550f625e82f0fe4e8abac369fb0c725a29d6/src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb#L175

**Comment by justinstoller** [comment]
Ugh, not sure if that's the bug in JRuby but it does look like that code in PAL should be supported. Looking into it.

**Comment by justinstoller** [comment]
No, I think that's an us problem. We're relying on pre Ruby 3 automatically converting hashes to kwargs. I think that's just a terrible message.

**Comment by justinstoller** [comment]
I wonder where else we're doing that and it just happens to work.....

**Comment by donoghuc** [comment]
OH, yeah i remember that. Its a ruby 3 compatability thing. I did it for bolt https://github.com/puppetlabs/bolt/pull/3093/commits/acda807a1c0c00f8221ebad8c7acd446dda8ae4f

**Comment by justinstoller** [comment]
Running the basic Puppet Server acceptance tests pass on Redhat8 with an Ubuntu22.04 agent:
```


              - Test Case Summary for suite 'tests' -
       Total Suite Time: 1176.56 seconds
      Average Test Time: 69.21 seconds
              Attempted: 17
                 Passed: 16
                 Failed: 0
                Errored: 0
                Skipped: 1
                Pending: 0
                  Total: 17
```

But they fail to start using CentOS 7.

The issue involves Puppet being unable to determine the OS type. It appears that the Ruby library function `Etc.getpwid(0)` (which is root) is returning `nil`. The `Etc` module in JRuby is re-implemented in Java and uses FFI bindings to call out to the OS, and if there's any issue then it returns nil.

The version of FFI being reported from JRuby 9.4 is the same as JRuby 9.3 though there are two possible culprits. One is that I believe JRuby vendors the libffi bindings within its jar. The other is that it now includes the libposixfix (via [subspawn](https://github.com/byteit101/subspawn#installation)) Both of these contain APIs to access the OS but, I believe, are compiled against a glibc that is too new for EL 7 and similar.

**Comment by justinstoller** [comment]
Aaand running the "full" acceptance tests (which are all the tests in the Puppet repo that have the "server" tag) was successful on Redhat 8:
```
- Test Case Summary for suite 'tests' -
       Total Suite Time: 2000.53 seconds
      Average Test Time: 8.70 seconds
              Attempted: 230
                 Passed: 54
                 Failed: 0
                Errored: 0
                Skipped: 176
                Pending: 0
                  Total: 230
```

**Comment by justinstoller** [comment]
WIP changes to get this working on CentOS 7: https://github.com/puppetlabs/jruby-deps/pull/43

**Comment by justinstoller** [comment]
Let me do some more manual testing to verify on older Ubuntus and newer boxes with this build.

**Comment by justinstoller** [comment]
One EL 8:
```
[root@decent-rocking ~]# cat /etc/os-release 
NAME="Red Hat Enterprise Linux"
VERSION="8.0 (Ootpa)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="8.0"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Red Hat Enterprise Linux 8.0 (Ootpa)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8.0:GA"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_BUGZILLA_PRODUCT_VERSION=8.0
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.0"

[root@decent-rocking ~]# puppetserver --version
puppetserver version: 7.9.4.SNAPSHOT.2023.01.30T1427

[root@decent-rocking ~]# puppetserver irb

irb(main):001:0> require 'pty'
=> true

irb(main):002:0> m, s = PTY.open
=> [#<IO:masterpty:/dev/pts/1>, #<File:/dev/pts/1>]

irb(main):003:0> r, w = IO.pipe
=> [#<IO:fd 15>, #<IO:fd 17>]

irb(main):004:0> pid = spawn("facter", in: r, out: s)
=> 25339

irb(main):005:0> m.gets
=> "aio_agent_version => 7.21.0.99\r\n"
```

**Comment by justinstoller** [comment]
Manual testing so far on SLES 15 for PTY support so far looks good:
```
frailest-gate:~ # cat /etc/os-release
NAME="SLES"
VERSION="15"
VERSION_ID="15"
PRETTY_NAME="SUSE Linux Enterprise Server 15"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15"

frailest-gate:~ # puppetserver --version
puppetserver version: 7.9.4.SNAPSHOT.2023.01.30T1427

frailest-gate:~ # puppetserver irb

irb(main):001:0' require 'pty'
=> true

irb(main):002:0> m, s = PTY.open
=> [#<IO:masterpty:/dev/pts/1>, #<File:/dev/pts/1>]

irb(main):003:0> r, w = IO.pipe
=> [#<IO:fd 22>, #<IO:fd 24>]

irb(main):004:0> pid = spawn("facter", in: r, out: s)
=> 30208

irb(main):005:0> m.gets
=> "aio_agent_version => 7.21.0.99\r\n"
```
However it fails on two of the smoke tests:
acceptance/suites/tests/https/client_may_use_external_cert_chains.rb
acceptance/suites/tests/010-puppetserver-cli/subcommand/gem.rb

The gem test failures is because installing nokogiri runs against the 1G memory limit set. I still have to triage the external_cert_chain_issue.

**Comment by justinstoller** [comment]
SLES 12 failed with the same cert issue while ubuntu 18 & 20 passed.

**Comment by justinstoller** [comment]
I should clarify that the centos7 & redhat8 runs where both the "smoke" and "full" acceptance while the sles, ubuntu, and fips tests were using just the smoke tests.

The "smoke" are the tests inside this repo that largely test JRuby integration & CA functionality - functionality that is unique to this repo - while the "full" tests are tests inside the puppet repo that are tagged with "server", ie, all the old functionality of the puppet master.

The main question we have right now is around the jni native functionality which will be tested by the smoke tests. Issues with Ruby code vs JRuby 9.4 should be largely the same amongst hosts.

**Comment by justinstoller** [comment]
Continuing to look into the SLES client using an external CA test.

**Comment by justinstoller** [comment]
Okay, messing with SLES I remember what's going on.

I made a comment last year regarding it.
https://github.com/puppetlabs/ci-job-configs/blob/main/jenkii/platform/projects/puppetserver.yaml#L264-L274

tl;dr - we no longer test SLES in FOSS because things we need are not available (it specifically calls out the Java version but when trying to manually replicate the issues openssl was too old to use the commands in the test as well).

I think this is the reason behind the SLES failures in 12 & 15 and is fine to ignore. We will have testing in PE for these platforms where we use our own Java & OpenSSL.

**Comment by justinstoller** [comment]
FIPS smoke tests finished and redhatfips 7 passed, while redhatfips 8 failed on one test. The test failed because we're trying to load non-fips key factories. This makes sense because we don't support support fips in FOSS, but it shows we don't have any additional libc and/or compiled library issues.

**Comment by justinstoller** [comment]
Summarizing the last two days of results:
 * The newest PR to jruby-deps resolves issues with older OSes.
 * Full acceptance passes on CentOS 7 & RedHat 8
 * We've smoke tested all platforms:
   * FIPS passes where expected w/o properly building this into PE
   * SLES 12 & 15 fail on a test they are currently failing on in main & 7.x, and it shouldn't fail once it's fully integrated in PE
   * SLES 15 fails the gem install test as it uses more memory than the process is allowed (installing nokogiri):
```
         Error: Your application used more memory than the automatic cap of 1324MB.
         Specify -J-Xmx####M to increase it (#### = cap size in MB).
         Specify -w for full java.lang.OutOfMemoryError: GC overhead limit exceeded stack trace>
```

I've verified that this isn't an issue prior to the patch. Not sure what the best approach is for this gem install issue (also not sure f it's that big of a deal). Will revisit in the morning.

**Comment by justinstoller** [comment]
Okay, I was able to verify that the `puppetserver gem` command only OOMs using SLES 15's old, custom patched Java 8 build. It does behave that way when using SLES 15's default java (which weirdly is Java 10). Since Java 8 support is being removed in Puppet Server 8 I think this is a non-issue. 

For users who want to stick with Java 8 anyways they have the ability to adjust the memory settins via the `JAVA_ARGS_CLI` environment variable, ie `JAVA_ARGS_CLI="-Xmx2000M" /opt/puppetserver/bin/puppetserver gem install nokogiri`

**Comment by justinstoller** [comment]
The final task for this PR is to enumerate why the JRuby maintainers believe the thread shutdown issue to be resolved yet our thread shutdown test fails and the server seems to leak threads in our integration tests (without the Timeout library kludge included in this PR).

File: project.clj
Comment (donoghuc) [inline project.clj:54]:
```suggestion
                 [puppetlabs/jruby-deps "9.4.0.0-2"]
```
long live `9.4.0.0-2` !  https://github.com/puppetlabs/jruby-deps/releases/tag/9.4.0.0-2

File: src/ruby/puppetserver-lib/puppet/server/master.rb
Comment (donoghuc) [inline src/ruby/puppetserver-lib/puppet/server/master.rb:166]: :smiling-face-but-you-can-see-the-pain-in-their-eyes:

**Review by donoghuc — APPROVED** [review]
Nice work on this! Hopefully we can get to the bottom of the that pesky timeout thread.

**Comment by justinstoller** [comment]
I was able to get to the bottom of the JRuby issue and a fix is up here: https://github.com/jruby/jruby/pull/7633 we're going to hold off on merging this until we get a feel for when a version of JRuby with that fix may be released.

**Review by donoghuc — APPROVED** [review]

**Pull request merged**
